processor.go

```diff
@@ -80,36 +83,45 @@ func (pro *Processor) OnProposal(proposal *message.Proposal) error {
 	// 1) try to confirm the parent vertex of the proposal
 	err := pro.confirmParent(proposal)
 	if err != nil {
+		// this error contains information about the reason for non-confirmation and is eventually handled in the caller scope
 		return rich.Errorf("could not confirm parent: %w", err)
 	}
 
+	// parent is confirmed -> move to applying the candidate message
+
 	// 2) try to apply the candidate vertex of the proposal
 	err = pro.applyCandidate(proposal)
 	if err != nil {
+		// this error contains information about the reason for not applying candidate and is eventually handled in the caller scope
 		return rich.Errorf("could not apply candidate: %w", err)
 	}
 
 	// 3) extract the proposer vote from the proposal (if we are collector)
 	err = pro.extractVote(proposal)
+	// candidate is successfully applied at this point -> so extract vote
```

Comment:
Why do we extract the vote?

Reply:
Extracting the vote is needed to check if the node that receives the proposal is the collector for it.
If it is - the peer needs to vote for the proposal and forward it so the next leader or proposer will receive it.
In this way the peer includes its vote on the proposal and gets its rewards if it is accepted.
As explained in https://github.com/awfm9/trickle/pull/1#discussion_r777381008, it is also pushing forward the voting process in case when a single node vote is needed to reach a majority (and we dont know that so we do it anyway)